Βασίλειος Μυλωνάς t8220103

In [37]:
import pandas as pd
import pingouin as pg
import matplotlib.pyplot as plt
import seaborn as sns


Task 1

I got the file i need from the paper's repo

In [30]:
gpt_dim = pd.read_csv('GPT2-large-dimensions.csv')

In [31]:
gpt_dim

Social.Category Source  gender.main  gender_norm.main  gender.ext  \
0       abandoned infant   GPT2    -0.197087          0.540877    0.001662   
1       abandoned person   GPT2    -0.199134          0.535577    0.001236   
2             abiogenist   GPT2    -0.231156          0.452671    0.004317   
3     able-bodied seaman   GPT2    -0.162948          0.629264    0.001341   
4            able seaman   GPT2    -0.181426          0.581424    0.002330   
...                  ...    ...          ...               ...         ...   
3490         young woman   GPT2    -0.156219          0.646685    0.016594   
3491           youngster   GPT2    -0.254537          0.392137   -0.005031   
3492        zen buddhist   GPT2    -0.242310          0.423792   -0.014386   
3493             zionist   GPT2    -0.239291          0.431610   -0.004601   
3494          zoo keeper   GPT2    -0.220986          0.479001    0.004355   

      gender_norm.ext  gender.red  gender_norm.red  age.main  age_norm.main  \
0            0.527162   -0.108243         0.561858 -0.070237       0.357426   
1            0.523567   -0.110507         0.554830 -0.065570       0.382294   
2            0.549545   -0.151075         0.428889 -0.097012       0.214761   
3            0.524456   -0.067956         0.686928 -0.042036       0.507690   
4            0.532793   -0.089043         0.621465 -0.055517       0.435856   
...               ...         ...              ...       ...            ...   
3490         0.653068   -0.067172         0.689364 -0.046145       0.485793   
3491         0.470725   -0.160560         0.399443 -0.098494       0.206867   
3492         0.391841   -0.146272         0.443798 -0.080021       0.305295   
3493         0.474350   -0.151177         0.428572 -0.092900       0.236669   
3494         0.549869   -0.136578         0.473894 -0.088563       0.259778   

       age.ext  age_norm.ext   age.red  age_norm.red  
0    -0.004732      0.319985 -0.055446      0.419423  
1    -0.001178      0.343708 -0.048989      0.477159  
2    -0.026444      0.175097 -0.063784      0.344863  
3     0.016286      0.460250 -0.043687      0.524568  
4     0.006879      0.397472 -0.049394      0.473540  
...        ...           ...       ...           ...  
3490  0.007785      0.403518 -0.038544      0.570562  
3491 -0.019306      0.222728 -0.073979      0.253696  
3492 -0.004964      0.318440 -0.063589      0.346609  
3493 -0.019493      0.221480 -0.066931      0.316723  
3494 -0.021305      0.209386 -0.059555      0.382674  

[3495 rows x 14 columns]

Here i change the column names so that they can be analyzed with pandas

In [32]:
gpt_dim = gpt_dim.rename(columns={ "gender.main": "gender_score",
                         "gender_norm.main": "gender_norm_score",
                         "gender.ext": "gender_ext",
                         "gender_norm.ext": "gender_norm_ext",
                         "gender.red": "gender_red",
                         "gender_norm.red": "gender_norm_red",
                         "age.main": "age_score",
                         "age_norm.main": "age_norm_score",
                         "age.ext": "age_ext",
                         "age_norm.ext": "age_norm_ext",
                         "age.red": "age_red",
                         "age_norm.red": "age_norm_red",
                         "Social.Category": "Social_Category"
                        
                         })

In [34]:
tab = pg.corr(
    gpt_dim["gender_score"],
    gpt_dim["age_score"]
)
tab

n         r         CI95%  p-val BF10  power
pearson  3495  0.872559  [0.86, 0.88]    0.0  inf    1.0

based on this discussion in stackoverflow, i found that the html table in the assignment is produced from _repr_html() https://stackoverflow.com/questions/27139896/ipython-repr-html

In [28]:
html = " ".join(tab._repr_html_().split())
html


'<div> <style scoped> .dataframe tbody tr th:only-of-type { vertical-align: middle; } .dataframe tbody tr th { vertical-align: top; } .dataframe thead th { text-align: right; } </style> <table border="1" class="dataframe"> <thead> <tr style="text-align: right;"> <th></th> <th>n</th> <th>r</th> <th>CI95%</th> <th>p-val</th> <th>BF10</th> <th>power</th> </tr> </thead> <tbody> <tr> <th>pearson</th> <td>3495</td> <td>0.872559</td> <td>[0.86, 0.88]</td> <td>0.0</td> <td>inf</td> <td>1.0</td> </tr> </tbody> </table> </div>'

then copy the html output in a markdown cell

<div> <style scoped> .dataframe tbody tr th:only-of-type { vertical-align: middle; } .dataframe tbody tr th { vertical-align: top; } .dataframe thead th { text-align: right; } </style> <table border="1" class="dataframe"> <thead> <tr style="text-align: right;"> <th></th> <th>n</th> <th>r</th> <th>CI95%</th> <th>p-val</th> <th>BF10</th> <th>power</th> </tr> </thead> <tbody> <tr> <th>pearson</th> <td>3495</td> <td>0.872559</td> <td>[0.86, 0.88]</td> <td>0.0</td> <td>inf</td> <td>1.0</td> </tr> </tbody> </table> </div>

Task 2

In [123]:
gender = gpt_dim[['gender_red', 'gender_ext', 'gender_score']]
gender

gender_red  gender_ext  gender_score
0      -0.108243    0.001662     -0.197087
1      -0.110507    0.001236     -0.199134
2      -0.151075    0.004317     -0.231156
3      -0.067956    0.001341     -0.162948
4      -0.089043    0.002330     -0.181426
...          ...         ...           ...
3490   -0.067172    0.016594     -0.156219
3491   -0.160560   -0.005031     -0.254537
3492   -0.146272   -0.014386     -0.242310
3493   -0.151177   -0.004601     -0.239291
3494   -0.136578    0.004355     -0.220986

[3495 rows x 3 columns]

In [36]:
age = gpt_dim[['age_red', 'age_ext', 'age_score']]
age

age_red   age_ext  age_score
0    -0.055446 -0.004732  -0.070237
1    -0.048989 -0.001178  -0.065570
2    -0.063784 -0.026444  -0.097012
3    -0.043687  0.016286  -0.042036
4    -0.049394  0.006879  -0.055517
...        ...       ...        ...
3490 -0.038544  0.007785  -0.046145
3491 -0.073979 -0.019306  -0.098494
3492 -0.063589 -0.004964  -0.080021
3493 -0.066931 -0.019493  -0.092900
3494 -0.059555 -0.021305  -0.088563

[3495 rows x 3 columns]

After searching the web, i decided to use seaborn to create the heatmap and i found the colormap used in the assignment to be coolwarm, but with center = 0.

https://seaborn.pydata.org/generated/seaborn.heatmap.html

https://www.practicalpythonfordatascience.com/ap_seaborn_palette

In [157]:
co_gender = gender.corr()

sns.heatmap(co_gender, cmap="coolwarm", annot=True, center = 0, linewidth=.5)
plt.title("Correlation Heatmap")
plt.savefig("corr_heat_gender.svg", format="svg", bbox_inches = 'tight', pad_inches = 0.3)
plt.close()

In [158]:
co_age = age.corr()

sns.heatmap(co_age, cmap="coolwarm", annot=True, center = 0, linewidth=.5)
plt.title("Correlation Heatmap")
plt.savefig("corr_heat_age.svg", format="svg", bbox_inches = 'tight', pad_inches = 0.3)
plt.close()

Then get the pictures and show them

<img src="corr_heat_age.svg" width="300"/><img src="corr_heat_gender.svg" width="300"/>

Task 3

Based on intro_to_stats_using_python

In [159]:
import statsmodels.formula.api as smf

In [163]:
ols_mod = smf.ols("age_norm_score ~ gender_norm_score", data=gpt_dim)
ols_res = ols_mod.fit()
ols_table = ols_res.summary()

i found this discussion https://stackoverflow.com/questions/51734180/converting-statsmodels-summary-object-to-pandas-dataframe

In [180]:
ols_table_html = ols_table.tables[0].as_html() + ols_table.tables[1].as_html() + ols_table.tables[2].as_html()
ols_table_html = ols_table_html.replace("\n", "")
ols_table_html

'<table class="simpletable"><caption>OLS Regression Results</caption><tr>  <th>Dep. Variable:</th>     <td>age_norm_score</td>  <th>  R-squared:         </th>  <td>   0.761</td> </tr><tr>  <th>Model:</th>                   <td>OLS</td>       <th>  Adj. R-squared:    </th>  <td>   0.761</td> </tr><tr>  <th>Method:</th>             <td>Least Squares</td>  <th>  F-statistic:       </th>  <td>1.114e+04</td></tr><tr>  <th>Date:</th>             <td>Sat, 13 Dec 2025</td> <th>  Prob (F-statistic):</th>   <td>  0.00</td>  </tr><tr>  <th>Time:</th>                 <td>18:57:39</td>     <th>  Log-Likelihood:    </th>  <td>  5425.6</td> </tr><tr>  <th>No. Observations:</th>      <td>  3495</td>      <th>  AIC:               </th> <td>-1.085e+04</td></tr><tr>  <th>Df Residuals:</th>          <td>  3493</td>      <th>  BIC:               </th> <td>-1.083e+04</td></tr><tr>  <th>Df Model:</th>              <td>     1</td>      <th>                     </th>      <td> </td>    </tr><tr>  <th>Covarianc

then copy the html output in a markdown cell

<table class="simpletable"><caption>OLS Regression Results</caption><tr>  <th>Dep. Variable:</th>     <td>age_norm_score</td>  <th>  R-squared:         </th>  <td>   0.761</td> </tr><tr>  <th>Model:</th>                   <td>OLS</td>       <th>  Adj. R-squared:    </th>  <td>   0.761</td> </tr><tr>  <th>Method:</th>             <td>Least Squares</td>  <th>  F-statistic:       </th>  <td>1.114e+04</td></tr><tr>  <th>Date:</th>             <td>Sat, 13 Dec 2025</td> <th>  Prob (F-statistic):</th>   <td>  0.00</td>  </tr><tr>  <th>Time:</th>                 <td>18:57:39</td>     <th>  Log-Likelihood:    </th>  <td>  5425.6</td> </tr><tr>  <th>No. Observations:</th>      <td>  3495</td>      <th>  AIC:               </th> <td>-1.085e+04</td></tr><tr>  <th>Df Residuals:</th>          <td>  3493</td>      <th>  BIC:               </th> <td>-1.083e+04</td></tr><tr>  <th>Df Model:</th>              <td>     1</td>      <th>                     </th>      <td> </td>    </tr><tr>  <th>Covariance Type:</th>      <td>nonrobust</td>    <th>                     </th>      <td> </td>    </tr></table><table class="simpletable"><tr>          <td></td>             <th>coef</th>     <th>std err</th>      <th>t</th>      <th>P>|t|</th>  <th>[0.025</th>    <th>0.975]</th>  </tr><tr>  <th>Intercept</th>         <td>   -0.0211</td> <td>    0.003</td> <td>   -6.510</td> <td> 0.000</td> <td>   -0.027</td> <td>   -0.015</td></tr><tr>  <th>gender_norm_score</th> <td>    0.7450</td> <td>    0.007</td> <td>  105.565</td> <td> 0.000</td> <td>    0.731</td> <td>    0.759</td></tr></table><table class="simpletable"><tr>  <th>Omnibus:</th>       <td>829.901</td> <th>  Durbin-Watson:     </th> <td>   1.255</td></tr><tr>  <th>Prob(Omnibus):</th> <td> 0.000</td>  <th>  Jarque-Bera (JB):  </th> <td>4643.267</td></tr><tr>  <th>Skew:</th>          <td> 1.012</td>  <th>  Prob(JB):          </th> <td>    0.00</td></tr><tr>  <th>Kurtosis:</th>      <td> 8.271</td>  <th>  Cond. No.          </th> <td>    9.75</td></tr></table>

Task 4

In [247]:
male_labels = [
    "chairman of the board",
    "elected official",
    "director of research",
    "chief of staff",
    "military personnel",
]


In [223]:
female_labels = [
    "intern",
    "novice",
    "cook",
    "secretary",
    "homoeopath"
]


In [252]:
male_df = gpt_dim[gpt_dim["Social_Category"].isin(male_labels)]



In [254]:
female_df = gpt_dim[gpt_dim["Social_Category"].isin(female_labels)]


In [ ]:
rest_df = gpt_dim[~gpt_dim["Social_Category"].isin(male_labels)]
rest_df = rest_df[~rest_df["Social_Category"].isin(female_labels)]


In [ ]:
from plotnine import *
%matplotlib inline

In [281]:
male_df

Social_Category Source  gender_score  gender_norm_score  \
533   chairman of the board   GPT2     -0.033962           0.963212   
577          chief of staff   GPT2     -0.083152           0.835858   
1034   director of research   GPT2     -0.019753           1.000000   
1154       elected official   GPT2     -0.072702           0.862913   
2360     military personnel   GPT2     -0.096926           0.800198   

      gender_ext  gender_norm_ext  gender_red  gender_norm_red  age_score  \
533     0.050124         0.935809    0.019927         0.959761  -0.011511   
577     0.038416         0.837081   -0.021661         0.830652  -0.033876   
1034    0.057737         1.000000    0.032889         1.000000  -0.015162   
1154    0.026189         0.733981   -0.002932         0.888795  -0.016559   
2360    0.009372         0.592171   -0.015763         0.848963  -0.013708   

      age_norm_score   age_ext  age_norm_ext   age_red  age_norm_red  
533         0.670336  0.013090      0.438922 -0.002641      0.891603  
577         0.551170  0.008316      0.407063 -0.023769      0.702679  
1034        0.650883  0.010761      0.423383 -0.006746      0.854894  
1154        0.643438  0.023949      0.511389 -0.018447      0.750263  
2360        0.658630  0.036275      0.593646 -0.028089      0.664048

In [314]:
annot = pd.DataFrame([
    # ---- MALE-CODED ----
    {
        "label": "chairman of the board",
        "lx": 1.00,
        "ly": 0.90,
        "color": "blue",
    },
    {
        "label": "director of research",
        "lx": 1.11,
        "ly": 0.55,
        "color": "blue",
    },
    {
        "label": "chief of staff",
        "lx": 0.88,
        "ly": 0.40,
        "color": "blue",
    },
    {
        "label": "military personnel",
        "lx": 0.50,
        "ly": 0.69,
        "color": "blue",
    },
    {
        "label": "elected official",
        "lx": 0.80,
        "ly": 0.80,
        "color": "blue",
    },

    # ---- FEMALE-CODED ----
    {
        "label": "intern",
        "lx": -0.06,
        "ly": 0.10,
        "color": "orange",
    },
    {
        "label": "cook",
        "lx": 0.10,
        "ly": 0,
        "color": "orange",
    },
    {
        "label": "secretary",
        "lx": 0.40,
        "ly": 0,
        "color": "orange",
    },
    {
        "label": "novice",
        "lx": 0.5,
        "ly": 0.05,
        "color": "orange",
    },
    {
        "label": "homoeopath",
        "lx": 0,
        "ly": 0.2,
        "color": "orange",
    }
])

In [315]:
annot = annot.merge(
    gpt_dim[[
        "Social_Category",
        "gender_norm_score",
        "age_norm_score"
    ]],
    left_on="label",
    right_on="Social_Category",
    how="left"
)


In [316]:
annot = annot.rename(columns={
    "gender_norm_score": "x",
    "age_norm_score": "y"
})


In [317]:
annot

label    lx    ly   color        Social_Category         x  \
0  chairman of the board  1.00  0.90    blue  chairman of the board  0.963212   
1   director of research  1.11  0.55    blue   director of research  1.000000   
2         chief of staff  0.88  0.40    blue         chief of staff  0.835858   
3     military personnel  0.50  0.69    blue     military personnel  0.800198   
4       elected official  0.80  0.80    blue       elected official  0.862913   
5                 intern -0.06  0.10  orange                 intern  0.014183   
6                   cook  0.10  0.00  orange                   cook  0.165446   
7              secretary  0.40  0.00  orange              secretary  0.306086   
8                 novice  0.50  0.05  orange                 novice  0.313752   
9             homoeopath  0.00  0.20  orange             homoeopath  0.136193   

          y  
0  0.670336  
1  0.650883  
2  0.551170  
3  0.658630  
4  0.643438  
5  0.056865  
6  0.072551  
7  0.096243  
8  0.116941  
9  0.116632

Created using plotnine.

Started from a basic scatter + lm fit. Axis limits are fixed and the fit is
extended manually to the full range. Background points are hollow to avoid
overplotting.

Highlighted points are handled separately. Label positions are defined
manually (lx / ly) in a small annotation table, merged back to the data,
and connected to points with line segments. Colors are passed through an
identity scale to avoid ggplot’s default remapping.


In [320]:
p = (
    ggplot(gpt_dim, aes(x="gender_norm_score", y="age_norm_score"))

    # background cloud
   + geom_point(
    data=rest_df,
    mapping=aes(x="gender_norm_score", y="age_norm_score"),
    size=0.8,
    fill="none",
    color="black",
    stroke=0.6,
    shape="o"
)


    # regression
    + geom_smooth(method="lm", se=False, color="red", fullrange=True)

    + geom_point(
    data=annot,
    mapping=aes(
        x="x",
        y="y",
        fill="color"  # column name
    ),
    # fill="color",
    size=3.5,
    color="black",
    stroke=0.6,
    shape="o"
)



    # connector lines
    + geom_segment(
        data=annot,
        mapping=aes(
            x="x", y="y",
            xend="lx", yend="ly"
        ),
        size=0.2,
        color="black"
    )

    # labels (ALL centered, as you want)
    + geom_text(
        data=annot,
        mapping=aes(
            x="lx", y="ly",
            label="label"
        ),
        ha="center",
        va="center",
        size=10
    )
    + scale_fill_identity()

    # axes
    + labs(
        x="Gender Association\n(Female–Male Dimension)",
        y="Age Association\n(Young–Old Dimension)"
    )
    + scale_x_continuous(limits=(-0.25, 1.35), expand=(0, 0))
    + scale_y_continuous(limits=(-0.20, 1.20), expand=(0, 0))
)
p.save(
    "age_gender_plot.svg",
    width=10,
    height=8,
    units="in"
)



c:\Users\Βασιλης\Applied_ML\.venv\Lib\site-packages\plotnine\ggplot.py:623: PlotnineWarning: Saving 10 x 8 in image.
c:\Users\Βασιλης\Applied_ML\.venv\Lib\site-packages\plotnine\ggplot.py:624: PlotnineWarning: Filename: age_gender_plot.svg
c:\Users\Βασιλης\Applied_ML\.venv\Lib\site-packages\plotnine\layer.py:374: PlotnineWarning: geom_smooth : Removed 1 rows containing missing values.


<img src="age_gender_plot.svg"/>

lets make this interactive

In [323]:
import altair as alt


In [352]:
base = alt.Chart(rest_df).mark_point(
    fill='lightgray',
    stroke='black',
    strokeWidth=0.1,
    size=40
).encode(
    x=alt.X(
        'gender_norm_score:Q',
        title='Gender Association (Female–Male Dimension)'
    ),
    y=alt.Y(
        'age_norm_score:Q',
        title='Age Association (Young–Old Dimension)'
    ),
    tooltip=[
        'Social_Category:N',
        alt.Tooltip('gender_norm_score:Q', format='.3f'),
        alt.Tooltip('age_norm_score:Q', format='.3f')
    ]
)
reg_line = alt.Chart(gpt_dim).transform_regression(
    'gender_norm_score',
    'age_norm_score'
).mark_line(
    color='red',
    strokeWidth=2
).encode(
    x='gender_norm_score:Q',
    y='age_norm_score:Q'
)
highlight_pts = alt.Chart(annot).mark_point(
    filled=True,
    stroke='black',
    strokeWidth=0.2,
    size=120
).encode(
    x='x:Q',
    y='y:Q',
    fill=alt.Fill('color:N', scale=None),
    tooltip=['label:N', 'x:Q', 'y:Q']
)

hover = alt.selection_point(
    fields=['label'],
    on='mouseover',
    empty=False
)
highlight_pts = highlight_pts.add_params(hover).encode(
    opacity=alt.condition(hover, alt.value(1), alt.value(0.7))
)

chart = (
    base
    + reg_line
    + highlight_pts
).properties(
    width=700,
    height=550
)
chart

alt.LayerChart(...)

In [353]:
chart.save("age_gender_interactive.html")